packages/core/core-flows/src/auth/workflows/generate-reset-password-token.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  AuthWorkflowEvents,
+  generateJwtToken,
+  MedusaError,
+} from "@medusajs/utils";
+import {
+  createWorkflow,
+  transform,
+  WorkflowResponse,
+} from "@medusajs/workflows-sdk";
+import { emitEventStep, useRemoteQueryStep } from "../../common";
+
+export const generateResetPasswordTokenWorkflow = createWorkflow(
+  "generate-reset-password-token",
+  (input: { entityId: string; provider: string }) => {
+    const providerIdentities = useRemoteQueryStep({
+      entry_point: "provider_identity",
+      fields: ["auth_identity_id", "provider_metadata"],
+      variables: {
+        filters: {
+          entity_id: input.entityId,
+          provider: input.provider,
+        },
+      },
+    })
+
+    const token = transform(
+      { input, providerIdentities },
+      ({ input, providerIdentities }) => {
+        const providerIdentity = providerIdentities?.[0]
+
+        if (!providerIdentity) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Provider identity with entity_id ${input.entityId} and provider ${input.provider} not found`
+          )
+        }
+
+        const token = generateJwtToken(
+          {
+            entity_id: input.entityId,
+            provider: input.provider,
+          },
+          {
+            // Ensures the token can only be used once per requested reset
```

Comment:
I think this opens up another set of attacks, which might be more concerning than someone being able to use the same token twice.

This basically makes the password hashing pointless. Imagine this scenario - You leak DB data on auth identities. Now, an attacker can now generate a JWT for any user, and reset their password, without having access to their email, nor knowing what their password is. This is equivalent to having the password in plaintext.

I would suggest we sign the token with the JWT secret - the reasons are:
1. Leaking DB data doesn't expose how your tokens are signed
2. In the case of leaking the JWT secret, rotating it will invalidate every single token.

I believe this results in a more secure setup.


Now, we can support token invalidation but I think for that it's better if we store that information in storage (eg. Redis with expiry)

Reply:
> Now, an attacker can now generate a JWT for any user, and reset their password, without having access to their email, nor knowing what their password is

The token would not be available in the Rest API response, so the only way to obtain the token would be through whatever subscription mechanism is setup, e.g. emails. 

If we are concerned about using the password hash, we could just use a substring of it, e.g. the first 6 characters.

Reply:
Having said that, I am fine using the JWT secret. This was my initial approach (if you look at previous commits), and is a very easy change to the PR. I am not expert on the subject matter, so if your experience tells you this is a better approach, we should do that.

Reply:
We discussed this in Slack a bit more, and decided to go with the JWT secret for now. We need to tackle invalidation for both invite links and standard tokens as a separate work.